### PR TITLE
 [C23] Add INFINITY and NAN macros to <float.h>

### PR DESCRIPTION
**Pull request by AaronBallman**
This is in support of WG14 N2848 which only define the macros if
infinity and nan are supported, so use of -ffinite-math will cause the
macros to not be defined.

### CODE REVIEW

**Comment by AaronBallman** [comment]
Link to WG14 N2848: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2848.pdf

**Review by efriedma-quic — COMMENTED** [review]
I'm not sure about tying this to `__FINITE_MATH_ONLY__`; -ffinite-math-only doesn't mean infinity doesn't exist, it just means you're promising that you won't use floating-point arithmetic/comparison ops on infinity.  Which is weird, but that's fast-math.  Also, other implementations don't do this.

In C99, INFINITY and NAN are macros in math.h, not float.h.  Is it correct to check `__STDC_VERSION__ >= 199901L`?

**Comment by AaronBallman** [comment]
> I'm not sure about tying this to `__FINITE_MATH_ONLY__`; -ffinite-math-only doesn't mean infinity doesn't exist, it just means you're promising that you won't use floating-point arithmetic/comparison ops on infinity. Which is weird, but that's fast-math. Also, other implementations don't do this.

The requirement that you only define the macro if inf/nan "is supported" is new in C23, so I don't expect anyone to have implemented this yet. As for `__FINITE_MATH_ONLY__`, our [documentation](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-ffinite-math-only) mention "results" and that implies (at least to me) that an expression cannot form an infinity to begin with, so the act of trying to expand `INFINITY` is nonsensical in that case, right?

> In C99, INFINITY and NAN are macros in math.h, not float.h. Is it correct to check __STDC_VERSION__ >= 199901L?

In C99, C11, and C17 `INFINITY` and `NAN` are only defined in `math.h`. In C23, the definition is now in both `math.h` and `float.h`. The `float.h` requirement comes from 5.2.5.3.3 `Characteristics of floating types <float.h>` (p29 and p3), and the `math.h` requirement comes from 7.12 `Mathematics <math.h>` (p7 and p9). I'm not 100% certain of why they're in both, but I think it's for freestanding conformance reasons (math.h isn't [always] in freestanding but float.h [always] is).

**Comment by efriedma-quic** [comment]
> and that implies (at least to me) that an expression cannot form an infinity to begin with, so the act of trying to expand INFINITY is nonsensical in that case, right?

It's undefined behavior at runtime.

I don't think we need to worry too much about what the C standard says here; the language around floating-point types that don't have infinity was written in the old days before everyone settled on using IEEE math.  And I think the warning we currently produce is probably more helpful in practice than "INFINITY not defined".

----

What I was getting at with the standard version thing is that in standard versions before C23, I don't think NAN is a reserved identifier if you don't include math.h.  So it's forbidden for float.h to define it.

**Comment by jcranmer-intel** [comment]
> I looked at my meeting notes for discussion of this paper and I think we do need to worry about what the C standard says. From my notes: `The big intent from this change seems to be about making INFINITY to be a feature test macro.`, so if users are going to portably expect `#ifdef INFINITY` to mean that their use of the macro will generate an actual infinity, we really should support that IMO.

I went looking through the CFP history of the proposal. It's based on a comp.std.c thread here: https://groups.google.com/g/comp.std.c/c/XCzEiwD9n_A, which became a CFP discussion here: https://mailman.oakapple.net/pipermail/cfp-interest/2021-October/002216.html (and more in the September portion of the archives, but the relevant part of the discussion is in October), and the minutes of the CFP meeting where this was discussed here: https://mailman.oakapple.net/pipermail/cfp-interest/2021-October/002253.html

The entire history of this discussion is about FP types that do not have a format for infinity (say, VAX FP), and there was commentary on the CFP thread that there was evidence that users thought that `INFINITY` was conditionally defined like `NAN` was. Note that `NAN` has been similarly conditionally defined for a long time, and gcc doesn't undefine it in `-ffinite-math-only` mode, so that's strong precedent for not undefining INFINITY in fast-math.

**Comment by AaronBallman** [comment]
> > I looked at my meeting notes for discussion of this paper and I think we do need to worry about what the C standard says. From my notes: `The big intent from this change seems to be about making INFINITY to be a feature test macro.`, so if users are going to portably expect `#ifdef INFINITY` to mean that their use of the macro will generate an actual infinity, we really should support that IMO.
> 
> I went looking through the CFP history of the proposal. It's based on a comp.std.c thread here: https://groups.google.com/g/comp.std.c/c/XCzEiwD9n_A, which became a CFP discussion here: https://mailman.oakapple.net/pipermail/cfp-interest/2021-October/002216.html (and more in the September portion of the archives, but the relevant part of the discussion is in October), and the minutes of the CFP meeting where this was discussed here: https://mailman.oakapple.net/pipermail/cfp-interest/2021-October/002253.html
> 
> The entire history of this discussion is about FP types that do not have a format for infinity (say, VAX FP), and there was commentary on the CFP thread that there was evidence that users thought that `INFINITY` was conditionally defined like `NAN` was. Note that `NAN` has been similarly conditionally defined for a long time, and gcc doesn't undefine it in `-ffinite-math-only` mode, so that's strong precedent for not undefining INFINITY in fast-math.

I spoke with the chair of CFP and he agreed with my assessment that `INFINITY` and `NAN` should not be defined in `-ffinite-math-only` mode, at least in terms of intent behind N2848. He also pointed out that not defining the macro is a security win because it makes it harder for people to write code using that macro and expecting it to do anything useful, which I strongly agree with.

Aside from "compatibility/historically", do we have any technical arguments against following the intent of the standard here? (We already deviate from GCC in so many other ways, I'm not worried about a deviation here; I would presume GCC will eventually get around to doing the same thing, at least for C23 and up).

**Comment by AaronBallman** [comment]
I was thinking about this more last night. The standard says:

> The macro
>    `INFINITY`
> is defined if and only if the implementation supports an infinity for the type float. It expands to a
> constant expression of type float representing positive or unsigned infinity.

Some interesting things to note:
* "if and only if" is clear that there are conditions under which `INFINITY` will not be defined.
* "an infinity" means some plausible, correct infinity value; because we `#pragma float_control(precise, on)`, there is a way in which to re-enable infinities for a block of code even if `-ffinite-math-only` is passed on the command line. That makes it defensible to always define `INFINITY`. However...
* "expands to a constant expression of type float representing positive or unsigned infinity" still prevents us from having a valid definition of the macro when infinity is disabled; we cannot make a valid constant expression because constant expressions are prohibited from overflowing, and because true infinity is considered representable but has properties that mean it doesn't overflow when used with mathematical operators. e.g., `INFINITY + 1.0f` is a valid constant expression only when true infinity is supported.
* You might think "but that's fine, we can add special handling for infinity to our constant expression interpreter", except there's a requirement for constant expressions that `The semantic rules for the evaluation of a constant expression are the same as for nonconstant expressions.`. It would also mean that `fpclassify` (etc) will classify such a value as an infinity at constant expression time, but not at runtime.
* We could also argue that `-ffinite-math-only` is a language dialect and so we could say "aha, none of the standard's rules apply!", but I think that's leaning on a technicality rather than helping users to write correct code. In general, we want language dialects to track as closely to the standard as they can because that's what makes the language dialect approachable for users.

I think that not defining `INFINITY` when in finite-math-only mode is the better option for the largest number of our users. People who want to play fast math games can call `__builtin_inf()` directly, but I think the fact that `INFINITY` won't actually behave like the standard expects it to behave and the standard asks us not to define it in that case is a compelling reason to not define it. I think that's the best user experience for people who aren't fast math aficionados (which is basically the majority of users, IMO).

WDYT?

**Comment by jcranmer-intel** [comment]
Some thoughts of my own:

The decision being discussed here has two main repercussions:
* An attempt to use `INFINITY` gets a symbol-not-found error message, or it gets whatever warning/error message we attach to `__builtin_inf()` in `-ffinite-math-only` mode.
* Code that uses `#ifndef INFINITY` may or may not trigger.

In the first case, I think you get a better error message with the `__builtin_inf()` usecase directly. For the second case, well, the main question is "why would someone use this in the first place." From a search on sourcegraph for existing uses of check for the `INFINITY` macro, by far the single most common case was to immediately define their own version of `INFINITY` by constructing an infinite value, usually by intentional overflow.

Most programmers, if they have any understanding of floating-point, would understand floating point only in terms of IEEE 754 types, which have an infinity; the existence of floating-point formats that lack infinities is going to be new to them, much less the ability to test such systems. So the lack of an `INFINITY` macro is likely to be seen less as "this compiler doesn't support infinities" and more likely to be seen as "this compiler doesn't support a new enough standard," hence the workarounds to construct an appropriate infinite value. (Many of the comments actually explicitly state that they are trying to construct an infinity!)

Code that attempts to intentionally use infinity in a `-ffinite-math-only` mode is going to break. The choice to me is giving them something where we will immediately complain that their combination of command-line flags and code is likely to do the wrong thing, or giving them something that--in practice--amounts to silently breaking their code, where we could have obviously told them their code was broken.

If the user is explicitly savvy about the existence of fast-math, they already have the opportunity to protect themselves from users who blindly apply `-ffast-math`, be it the `__FINITE_MATH_ONLY` macro checks or pragmas to re-enable it. Not defining `INFINITY` for their benefit doesn't give them a lot more control, while reducing our ability to help users who may have blundered themselves into a nonsense configuration.

In short: IMHO, `INFINITY` should expand to something that bops the user (i.e., warn/error by default) and says "you're doing something wrong" in finite-math-only mode.

**Comment by AaronBallman** [comment]
> In short: IMHO, `INFINITY` should expand to something that bops the user (i.e., warn/error by default) and says "you're doing something wrong" in finite-math-only mode.

@jcranmer-intel and I had a really nice offline discussion on this topic and I've come around to his view. I was thinking that the fact we don't currently diagnose use of `__builtin_inf()` in `-ffinite-math-only` mode was because it produced an actual infinity still, but after our discussion, I now agree that it (and `__builtin_nan()`) should diagnose if used when infinities (or NANs) are disabled. This also nicely handles the case with pragmas changing the floating-point mode to be different from what was set on the command line.

So let's split this PR into two parts: 1) I'll update this PR to always define `INFINITY` and `NAN` in `<float.h>`, 2) I'll introduce the warning diagnostic in a follow-up PR.

File: clang/lib/Headers/float.h [hunk 1]
```diff
@@ -47,6 +47,8 @@
     (defined(__cplusplus) && __cplusplus >= 201103L) ||                        \
     (__STDC_HOSTED__ && defined(_AIX) && defined(_ALL_SOURCE))
 #    undef DECIMAL_DIG
+#    undef INFINITY
```
Comment (efriedma-quic) [inline clang/lib/Headers/float.h:50]: The define guarding the undefs needs to be the same as the one for the defines, I think?

**Review by efriedma-quic — COMMENTED** [review]
We already have `-Wnan-infinity-disabled`, which I think is sufficient on the warning side?  Haven't checked carefully to see what cases it covers.

**Comment by AaronBallman** [comment]
> We already have `-Wnan-infinity-disabled`, which I think is sufficient on the warning side? Haven't checked carefully to see what cases it covers.

Not this one. :-D https://godbolt.org/z/hq8jvse9o But yeah, that may be a good diagnostic to consider putting this under, but I think that diagnostic should be enabled by default when passing -Wfinite-math-only or other flags that disable infinity or nan support.

File: clang/lib/Headers/float.h (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Headers/float.h:50]: Good catch, I've corrected this.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Comment by bgra8** [comment]
@AaronBallman [N2848](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2848.pdf) explicitly mentions that `NAN` and `INFINITY` macros should  expand to a constant expression of type `float` not `double` like was implemented in this patch.

`__builtin_nanf` and `__builtin_inff` should have been used instead, right?

**Comment by AaronBallman** [comment]
> @AaronBallman [N2848](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2848.pdf) explicitly mentions that `NAN` and `INFINITY` macros should expand to a constant expression of type `float` not `double` like was implemented in this patch.
> 
> `__builtin_nanf` and `__builtin_inff` should have been used instead, right?

WHOOOOOOPPPPSSS yup, looks like I also missed some test coverage. :-D I'll get that fixed up ASAP, thank you!

**Comment by AaronBallman** [comment]
> > @AaronBallman [N2848](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2848.pdf) explicitly mentions that `NAN` and `INFINITY` macros should expand to a constant expression of type `float` not `double` like was implemented in this patch.
> > `__builtin_nanf` and `__builtin_inff` should have been used instead, right?
> 
> WHOOOOOOPPPPSSS yup, looks like I also missed some test coverage. :-D I'll get that fixed up ASAP, thank you!

This should be fixed now in 2f2b931e1296aebe6c03fd969363683b637973e5, thank you!

**Comment by bgra8** [comment]
> > > @AaronBallman [N2848](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2848.pdf) explicitly mentions that `NAN` and `INFINITY` macros should expand to a constant expression of type `float` not `double` like was implemented in this patch.
> > > `__builtin_nanf` and `__builtin_inff` should have been used instead, right?
> > 
> > 
> > WHOOOOOOPPPPSSS yup, looks like I also missed some test coverage. :-D I'll get that fixed up ASAP, thank you!
> 
> This should be fixed now in [2f2b931](https://github.com/llvm/llvm-project/commit/2f2b931e1296aebe6c03fd969363683b637973e5), thank you!

Thanks a lot for the quick fix!
